# Loops (Iteration Structure)

### What are loops?
- A loop is an iterative structure
- It helps to execute a block (code) repeatedly
- There are two types of loops:
    - Collection (countable) loops - `for` loop
    - Conditional loops - `while` loop

### The `for` loop
- This is a collection loop
- For every item in the collection the block (code) executes once, and after that loop ends
- Syntax:

```
for <var> in <collection>:
    block
```

In [1]:
for x in [10, 18, -25]:
    print("Hello, World!")

Hello, World!
Hello, World!
Hello, World!


In [2]:
for i in [10, -100]:
    print("Hello, World!")
    print("Good evening!")

Hello, World!
Good evening!
Hello, World!
Good evening!


In [3]:
for i in ["Naveen", "Kumar", "Nagesh"]:
    print("Hello, " + i + "!")
    print("---------------")

Hello, Naveen!
---------------
Hello, Kumar!
---------------
Hello, Nagesh!
---------------


In [4]:
for x in [10, 18, 25, 44, 19]:
    print("Temp:", x)

Temp: 10
Temp: 18
Temp: 25
Temp: 44
Temp: 19


In [5]:
for p in (100, -200, 300, -400, -500, 600):
    print("Value is:", p)
    print("--------------")
    print("Looping is fun!")

Value is: 100
--------------
Looping is fun!
Value is: -200
--------------
Looping is fun!
Value is: 300
--------------
Looping is fun!
Value is: -400
--------------
Looping is fun!
Value is: -500
--------------
Looping is fun!
Value is: 600
--------------
Looping is fun!


In [6]:
for x in [1001, "Raju", 35, "Hyd", "HR", 48000.0, True]:
    print("Item:", x)

Item: 1001
Item: Raju
Item: 35
Item: Hyd
Item: HR
Item: 48000.0
Item: True


In [7]:
for l in "Help":
    print("Char:", l)

Char: H
Char: e
Char: l
Char: p


In [8]:
for i in ("Naveen", 35, "Kumar", 40):
    print("Name or Age:", i)

Name or Age: Naveen
Name or Age: 35
Name or Age: Kumar
Name or Age: 40


In [9]:
total = 0
for t in [10, 15, 20, 30, 35]:
    total = total + t
print("Total is:", total)

Total is: 110


In [10]:
total = 1
for i in [5, 2, 7, 4, 3, 9]:
	total = total * i
print("Final Total is:", total)

Final Total is: 7560


In [11]:
for i in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]:
    print("18 x", i, "=", 18 * i)

18 x 1 = 18
18 x 2 = 36
18 x 3 = 54
18 x 4 = 72
18 x 5 = 90
18 x 6 = 108
18 x 7 = 126
18 x 8 = 144
18 x 9 = 162
18 x 10 = 180


In [12]:
fact = 1
for i in [1, 2, 3, 4, 5]:
    fact = fact * i
print("Factorial is:", fact)

Factorial is: 120


In [13]:
for i in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]:
    print("18 x", i, "=", 18 * i)

18 x 1 = 18
18 x 2 = 36
18 x 3 = 54
18 x 4 = 72
18 x 5 = 90
18 x 6 = 108
18 x 7 = 126
18 x 8 = 144
18 x 9 = 162
18 x 10 = 180
18 x 11 = 198
18 x 12 = 216
18 x 13 = 234
18 x 14 = 252
18 x 15 = 270
18 x 16 = 288
18 x 17 = 306
18 x 18 = 324
18 x 19 = 342
18 x 20 = 360


#### a list of series
1. 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
2. 0, 1, 2, 3, 4, ...
3. -10, -9, -8, -7, ..., 0, 1, 2, 3, ..., 10
4. 4, 3, 2, 1, 0, -1, -2, -3, -4
5. 7, 10, 13, 16, 19, 22, 25, ...
6. 4, 1, -2, -5, -8, -11, ...
7. 8, 18, 28, 38, 48, 58, ..., 108
8. 5 x 2 ** 1, 5 x 2 ** 2, 5 x 2 ** 3, 5 x 2 ** 4, 5 x 2 ** 5, ....
9. 1/3, 1/5, 1/7, 1/9

### The `range( [start,] stop [, step] )` function/iterator
- The `range()` iterator has 3 paramters: start, stop, step
- One required parameter: stop
- Two optional parameters: start and step
- The default value of start is 0
- The default value of step is 1
- The *range()* iterator returns values from 0 (start value) till stop value, by continuously adding step value (1). The stop value is exclusive.

#### The `range(stop)`
- Now start is 0, and step is 1
- Examples:
    * range(5)      # 0, 1, 2, 3, 4
    * range(3)      # 0, 1, 2
    * range(8)      # 0, 1, 2, 3, 4, 5, 6, 7

In [14]:
for i in range(5):
    print(i, end=" ")

0 1 2 3 4 

In [15]:
for i in range(3):
    print(i, end=" ")

0 1 2 

In [16]:
for i in range(8):
    print(i, end=" ")

0 1 2 3 4 5 6 7 